tools/run_tests/xds_k8s_test_driver/framework/test_app/base_runner.py

```diff
@@ -35,14 +50,113 @@ class KubernetesBaseRunner:
+
+    def _log_container(self, core: client.CoreV1Api, pod_name: str,
+                       container_name: str, namespace: str) -> None:
+        query_restarted = False
+        logfile = os.path.join(
+            _TEST_LOG_BASE_DIR,
+            f"{self.case_name}.{pod_name}.{container_name}.sponge_log.log")
+
+        self._await_container_ready(core, pod_name, container_name, namespace)
+
+        with open(logfile, "w") as f:
+            while not self._log_stop_event.is_set():
+                try:
+                    if query_restarted:
+                        f.write(
+                            "Restarted log fetching. Attempting to read from the beginning, but truncation may have occurred.\n"
+                        )
+                    w = watch.Watch()
+                    for msg in w.stream(core.read_namespaced_pod_log,
+                                        name=pod_name,
+                                        namespace=namespace,
+                                        container=container_name,
+                                        follow=True):
+                        f.write(msg)
+                        f.write("\n")
+                except kubernetes.client.rest.ApiException as e:
+                    f.write(f"Exception fetching logs: {e}\n")
+                    query_restarted = True
+                    time.sleep(_BACKOFF_SECONDS)
+
+    def _start_logging_container(self, core: client.CoreV1Api, pod_name: str,
+                                 container_name: str, namespace: str) -> None:
+        t = threading.Thread(target=self._log_container,
+                             args=(core, pod_name, container_name, namespace),
+                             daemon=True)
+        t.start()
+
+    def _start_logging_pod(self, core: client.CoreV1Api, pod_name: str,
+                           namespace: str) -> None:
+        retryer = tenacity.Retrying(retry=tenacity.retry_if_exception_type(
+            kubernetes.client.rest.ApiException),
+                                    wait=tenacity.wait_fixed(_BACKOFF_SECONDS),
+                                    reraise=True)
+        pod = retryer(core.read_namespaced_pod, pod_name, namespace)
+        for container in pod.spec.containers:
+            self._start_logging_container(core, pod_name, container.name,
+                                          namespace)
+
+    def _start_logging_deployment(self, deployment_name: str):
+        core = self.k8s_namespace.api.core
+        apps = self.k8s_namespace.api.apps
+
+        pod_names = None
+        namespace = self.k8s_namespace.name
+
+        def _get_deployment_pods():
+            pods_all_namespaces = core.list_pod_for_all_namespaces(
+                label_selector=self.pod_label_selector).items
+            pods = [
+                pod for pod in pods_all_namespaces
+                if pod.metadata.namespace == namespace
+            ]
+            return [pod.metadata.name for pod in pods]
```

Comment:
nit: I think generator should work too
```suggestion
            return (pod.metadata.name for pod in pods)
```